From 1d4cc3813f8263bf089abdacacf65a5fcdb67650 Mon Sep 17 00:00:00 2001
From: Zhenxiao Luo <zhenxiao@cloudera.com>
Date: Wed, 27 Jun 2012 15:12:53 -0700
Subject: [PATCH 046/148] HIVE-2804. Task log retrieval fails on secure cluster, null check added

Reason: Bug
Author: Zhenxiao Luo
Ref: CDH-4488
---
 .../hadoop/hive/ql/exec/HadoopJobExecHelper.java   |    7 +++++--
 .../apache/hadoop/hive/ql/exec/JobDebugger.java    |    7 +++++--
 .../clientnegative/cluster_tasklog_retrieval.q.out |    2 +-
 3 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/src/ql/src/java/org/apache/hadoop/hive/ql/exec/HadoopJobExecHelper.java b/src/ql/src/java/org/apache/hadoop/hive/ql/exec/HadoopJobExecHelper.java
index a46499b..082d23c 100644
--- a/src/ql/src/java/org/apache/hadoop/hive/ql/exec/HadoopJobExecHelper.java
+++ b/src/ql/src/java/org/apache/hadoop/hive/ql/exec/HadoopJobExecHelper.java
@@ -540,8 +540,11 @@ public class HadoopJobExecHelper {
         }
         // These tasks should have come from the same job.
         assert (ti.getJobId() != null && ti.getJobId().equals(jobId));
-        ti.getLogUrls().add(ShimLoader.getHadoopShims().getTaskAttemptLogUrl(
-          conf, t.getTaskTrackerHttp(), t.getTaskId()));
+        String taskAttemptLogUrl = ShimLoader.getHadoopShims().getTaskAttemptLogUrl(
+          conf, t.getTaskTrackerHttp(), t.getTaskId());
+        if (taskAttemptLogUrl != null) {
+          ti.getLogUrls().add(taskAttemptLogUrl);
+        }
 
         // If a task failed, then keep track of the total number of failures
         // for that task (typically, a task gets re-run up to 4 times if it
diff --git a/src/ql/src/java/org/apache/hadoop/hive/ql/exec/JobDebugger.java b/src/ql/src/java/org/apache/hadoop/hive/ql/exec/JobDebugger.java
index 1a6db23..8fe76fa 100644
--- a/src/ql/src/java/org/apache/hadoop/hive/ql/exec/JobDebugger.java
+++ b/src/ql/src/java/org/apache/hadoop/hive/ql/exec/JobDebugger.java
@@ -145,8 +145,11 @@ public class JobDebugger implements Runnable {
           }
           // These tasks should have come from the same job.
           assert (ti.getJobId() != null &&  ti.getJobId().equals(jobId));
-          ti.getLogUrls().add(ShimLoader.getHadoopShims().getTaskAttemptLogUrl(
-            conf, t.getTaskTrackerHttp(), t.getTaskId()));
+          String taskAttemptLogUrl = ShimLoader.getHadoopShims().getTaskAttemptLogUrl(
+            conf, t.getTaskTrackerHttp(), t.getTaskId());
+          if (taskAttemptLogUrl != null) {
+            ti.getLogUrls().add(taskAttemptLogUrl);
+          }
 
           // If a task failed, then keep track of the total number of failures
           // for that task (typically, a task gets re-run up to 4 times if it
diff --git a/src/ql/src/test/results/clientnegative/cluster_tasklog_retrieval.q.out b/src/ql/src/test/results/clientnegative/cluster_tasklog_retrieval.q.out
index 3dfbeb7..457980a 100644
--- a/src/ql/src/test/results/clientnegative/cluster_tasklog_retrieval.q.out
+++ b/src/ql/src/test/results/clientnegative/cluster_tasklog_retrieval.q.out
@@ -10,5 +10,5 @@ PREHOOK: query: FROM src
 SELECT evaluate_npe(src.key) LIMIT 1
 PREHOOK: type: QUERY
 PREHOOK: Input: default@src
-PREHOOK: Output: file:/tmp/cloudera/hive_2012-05-17_16-25-32_591_2604050764036436885/-mr-10000
+#### A masked pattern was here ####
 FAILED: Execution Error, return code 2 from org.apache.hadoop.hive.ql.exec.MapRedTask
-- 
1.7.0.4

